src/mame/altos/altos586_hdc.cpp

```diff
@@ -0,0 +1,371 @@
+
+		// Sector header.
+		m_sector[2] = m_sector[0];
+		m_sector[0] = m_cyl[m_drive] & 0xff;
+		// Bit 3 indicates bad sector (we don't ever set it).
+		m_sector[1] = m_head << 4 | m_cyl[m_drive] >> 8;
+		m_sector[3] = m_sector[4] = 0;
+
+		sector_read(m_sector[2]);
+		m_secoffset = 0;
+		break;
+
+	case 0x02:
+		// Write.
+		// Sector number now in data buffer, data write follows.
+		LOG("WRITE command\n");
+		if (m_secoffset != 1)
+			logerror("expected one value in data buffer, has %d\n", m_secoffset);
+
+		// No sector format, just data.
+		m_secoffset = 5;
+		m_iop->drq1_w(ASSERT_LINE);
+		break;
+
+	case 0x04:
+		// Write Sector Format.
+		// Sector number now in data buffer, header write follows.
+		LOG("WRITE FORMAT command\n");
+		if (m_secoffset != 1)
+			logerror("expected one value in data buffer, has %d\n", m_secoffset);
+
+		m_secoffset = 1; // Leave the sector number in.
+		m_iop->drq1_w(ASSERT_LINE);
+		break;
+
+	case 0x10:
+		// Seek to cylinder.
+		// Current cylinder is in data buffer, new one in a separate latch.
+		LOG("SEEK from cylinder %d to %d\n", m_sector[1] << 8 | m_sector[0], m_cyl_latch);
+		if (m_secoffset != 2)
+			logerror("expected two values in data buffer, has %d\n", m_secoffset);
+
+		m_secoffset = 0;
+		m_cyl[m_drive] = m_cyl_latch;
+		// Seek done.
+		m_seek_status |= 0x02;
+		// Not busy.
+		m_status &= ~0x01;
+		break;
+
+	case 0x20:
+		// Not sure, actually.
+		LOG("SELECT command\n");
+		// Head/drive selected?
+		m_seek_status |= 0x01;
+		break;
+
+	case 0x80:
+		// This looks like a reset of some sort.
+		LOG("RESET command\n");
+		device_reset();
+		break;
+
+	default:
+		logerror("unknown command 0x%02x\n", data);
+	}
+}
+
+void altos586_hdc_device::altos586_hdc_io(address_map &map)
+{
+	map(0x0000, 0x3fff).ram(); // 16K SRAM.
+	map(0xffd0, 0xffd1).rw(FUNC(altos586_hdc_device::data_r), FUNC(altos586_hdc_device::data_w));
+	map(0xffd2, 0xffd3).w(FUNC(altos586_hdc_device::head_select_w));
+	map(0xffd4, 0xffd5).rw(FUNC(altos586_hdc_device::seek_status_r), FUNC(altos586_hdc_device::cyl_w));
+	map(0xffd6, 0xffd7).rw(FUNC(altos586_hdc_device::status_r), FUNC(altos586_hdc_device::command_w));
+	// Diags write 0x10 to 0xfff8 and XENIX writes 0x82 to 0xffe6.
+	// Not sure what either does.
+	map(0xfff8, 0xfff9).nopw();
+}
+
+std::error_condition altos586_hdc_device::hdd_load(device_image_interface &image, uint8_t index)
+{
+	if (m_hdd[index]->get_info().sectorbytes != 512) {
+		logerror("expected 512 bytes per sector, got %d", m_geom[index]->sectorbytes);
+		return image_error::INVALIDLENGTH;
+	}
+
+	m_geom[index] = &m_hdd[index]->get_info();
+	return std::error_condition();
+}
+
+void altos586_hdc_device::device_add_mconfig(machine_config &config)
+{
+	I8089(config, m_iop, XTAL(15'000'000) / 3);
+	m_iop->set_addrmap(AS_PROGRAM, &altos586_hdc_device::altos586_hdc_mem);
+	m_iop->set_addrmap(AS_IO, &altos586_hdc_device::altos586_hdc_io);
+	m_iop->set_data_width(16);
```

Comment:
Does this board have its own clock, or does it use the bus clock for this?  If it uses the bus clock, you can use the `DERIVED_CLOCK` helper for the 8089 clock frequency and set the bus clock when instantiating the controller board.